lib/filter.js

```diff
@@ -0,0 +1,141 @@
+  return {
+    status: "403",
+    code: "EFORBIDDEN",
+    title: "Invalid filter",
+    detail: "Filter relationship '" + key + "' is a foreign reference and does not exist on " + resourceConfig.resource
+  };
+};
+
+filter._splitElement = function(element) {
+  if (!element) return null;
+  if (FILTER_OPERATORS.indexOf(element[0]) !== -1) {
+    return { operator: element[0], value: element.substring(1) };
+  }
+  return { operator: null, value: element };
+};
+
+filter._stringOnlyOperator = function(operator, attributeConfig) {
+  if (!operator || !attributeConfig) return null;
+  if (STRING_ONLY_OPERATORS.indexOf(operator) !== -1 && attributeConfig._type !== "string") {
+    return "operator " + operator + " can only be applied to string attributes";
+  }
+  return null;
+};
+
+filter._parseScalarFilterElement = function(attributeConfig, scalarElement) {
+  if (!scalarElement) return { error: "invalid or empty filter element" };
+
+  var splitElement = filter._splitElement(scalarElement);
+  if (!splitElement) return { error: "empty filter" };
+
+  var error = filter._stringOnlyOperator(splitElement.operator, attributeConfig);
+  if (error) return { error: error };
+
+  if (attributeConfig._settings) {  // relationship attribute: no further validation
+    return { result: splitElement };
+  }
+
+  var validateResult = attributeConfig.validate(splitElement.value);
+  if (validateResult.error) {
+    return { error: validateResult.error.message };
+  }
+
+  var validatedElement = { operator: splitElement.operator, value: validateResult.value };
+  return { result: validatedElement };
+};
+
+filter._parseFilterElementHelper = function(attributeConfig, filterElement) {
+  if (!filterElement) return { error: "invalid or empty filter element" };
+
+  var parsedElements = [].concat(filterElement).map(function(scalarElement) {
+    return filter._parseScalarFilterElement(attributeConfig, scalarElement);
+  });
+
+  if (parsedElements.length === 1) return parsedElements[0];
+
+  var errors = parsedElements.reduce(function(combined, element) {
+    if (!combined) {
+      if (!element.error) return combined;
+      return [ element.error ];
+    }
+    return combined.concat(element.error);
+  }, null);
+
+  if (errors) return { error: errors };
+
+  var results = parsedElements.map(function(element) {
+    return element.result;
+  });
+
+  return { result: results };
+};
+
+filter._parseFilterElement = function(attributeName, attributeConfig, filterElement) {
+  var helperResult = filter._parseFilterElementHelper(attributeConfig, filterElement);
+
+  if (helperResult.error) {
+    return {
+      error: {
+        status: "403",
+        code: "EFORBIDDEN",
+        title: "Invalid filter",
+        detail: "Filter value for key '" + attributeName + "' is invalid: " + helperResult.error
+      }
+    };
+  }
+  return { result: helperResult.result };
+};
+
+filter.process = function(request, callback) {
```

Comment:
Is `filter.process` the right name for this? Reading further down and seeing it alongside other high level features left me a bit confused as to what processing it was doing - maybe something like `filter.validateAndSanitise` could be more appropriate?